### DISCUSSION TITLE
Rate Limit?

### DISCUSSION THREAD

**Discussion by 0x747** [dr5hn/countries-states-cities-database#919] [Q&A]
This [post](https://github.com/dr5hn/countries-states-cities-database/discussions/573) mentions limits. Anyone know what the actual limits/rates are? I couldn't find anything in the documentation.

**Comment by NDCallahan** [comment]
Yes, it's listed here: https://docs.countrystatecity.in/api/rate-limits The problem I have is for testing. These limits are going to limit testing these APIs by a lot..

**Comment by dr5hn** [comment]
@0x747 

Thanks for raising this! These rate limits are proposed - I'm gathering real-world usage data from the community before implementing anything.

@NDCallahan an Your testing concern is exactly what I need to understand better:

1. How many API calls does your typical testing workflow make per day?
2. What are you testing? (integration tests, data validation, development workflow?)
3. Have you considered using the database exports for testing? (https://export.countrystatecity.in)

I'm trying to find the right balance between:
- Keeping the API free and accessible
- Preventing abuse/heavy leeching that impacts performance
- Directing bulk usage to database exports (which are better for testing anyway)

**Real talk:** Most production apps use <100 requests/day. Testing and development workflows are where people hit thousands of calls - which is exactly why we offer full database exports at https://export.countrystatecity.in

What would work best for your use case?

**Reply by NDCallahan** [reply]
This is what I can say. For production, I'm probably not going to reach 100 calls per day. However, I am continuously developing and testing. 

This API is attached to user forms. Three different ones. So at any given time, I'll be lucky to hit 100 a day. Now, I'm testing I regression test everything. So, for example, today, I hit 65 calls. Not huge. Not tiny. 

The second thing I will say is this: If I use the export, that isn't a true testing senario for me if I'm actually using the API calls.

**Reply by NDCallahan** [reply]
Perhaps a header gets passed for development that is allowed to use a bigger set? The problem there is people would just use the dev header in production to bypass that probably. So lol. Maybe not a good idea.
